File: lib/features/chat/models/style.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'package:garbanzo_ai/features/chat/models/thinking_level.dart';

part 'style.freezed.dart';
part 'style.g.dart';

/// A saved style: a named bundle of model + thinking level + system prompt
/// template. Applying one configures a conversation in a single tap.
///
/// Built-in styles (isBuiltin) ship with the app, are shared across all
/// users, and are read-only — the backend refuses PATCH/DELETE on them.
/// `locale` carries a BCP-47 tag for built-ins so the picker surfaces them
/// in the user's language; null for user-saved styles (language-neutral).
@freezed
abstract class Style with _$Style {
  const factory Style({
    required String id,
    required String name,
    String? description,
    required String modelId,
    // null = provider default ("Auto"): thinking auto-enables when the
    // model supports it.
    @JsonKey(unknownEnumValue: JsonKey.nullForUndefinedEnumValue)
    ThinkingLevel? thinkingLevel,
    // References a SystemPromptTemplate; null = no persona. The backend
    // nulls this out (not cascades) when the template is deleted.
    String? systemPromptTemplateId,
    @Default(false) bool isBuiltin,
    String? locale,
    // Seeds new conversations. At most one style per user is default.
    @Default(false) bool isDefault,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = _Style;

  factory Style.fromJson(Map<String, dynamic> json) => _$StyleFromJson(json);
}

File: backend/app/schemas/style.py
"""Pydantic schemas for saved styles (Idea 2: "Styles")."""

from datetime import datetime

from pydantic import BaseModel, Field

from app.core.config import get_settings
from app.schemas.chat import ThinkingLevel


class StyleCreate(BaseModel):
    """Request to create a saved style."""

    name: str = Field(..., min_length=1, max_length=100)
    model_id: str = Field(
        default_factory=lambda: get_settings().default_model,
        max_length=100,
        description="The model this style selects",
    )
    thinking_level: ThinkingLevel | None = Field(
        None,
        description=(
            "Reasoning depth for thinking-capable models. None = provider "
            "default (auto-enable thinking when the model supports it)."
        ),
    )
    system_prompt_template_id: str | None = Field(
        None,
        description="A system prompt template this style applies, or null for none",
    )
    is_default: bool = Field(
        False,
        description=(
            "Make this the style used to seed new conversations. Setting "
            "this unsets any previous default for the user."
        ),
    )


class StyleUpdate(BaseModel):
    """Partial update for a saved style. Omitted fields are left unchanged.

    Built-in styles are read-only for content fields (name/model/thinking/
    template) — the endpoint returns 403 for those. ``is_default`` is the
    exception: it writes a per-user pointer, so a built-in can be set as
    the user's default for new chats.
    """

    name: str | None = Field(None, min_length=1, max_length=100)
    model_id: str | None = Field(None, max_length=100)
    thinking_level: ThinkingLevel | None = Field(
        None,
        description=(
            "Reasoning depth for thinking-capable models. Not sent in the "
            "payload -> unchanged. Sent as null -> reset to the provider "
            "default. Sent as 'off'/'low'/'medium'/'high' -> set that level."
        ),
    )
    system_prompt_template_id: str | None = Field(
        None,
        description=("Not sent in the payload -> unchanged. Sent as null -> clear it."),
    )
    is_default: bool | None = Field(
        None,
        description=(
            "Make/unmake this the default style for new conversations. "
            "Allowed for built-in styles too — it sets a per-user pointer "
            "rather than mutating the shared row."
        ),
    )


class StyleOut(BaseModel):
    """A saved style as returned by the API. Includes built-in flag,
    locale, and description so the picker can render built-ins consistently
    with user-saved styles and filter them by the user's language.
    """

    id: str
    name: str
    description: str | None = None
    model_id: str
    thinking_level: ThinkingLevel | None = None
    system_prompt_template_id: str | None = None
    is_builtin: bool = False
    locale: str | None = None
    is_default: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
